rclcpp/src/rclcpp/node_interfaces/node_base.cpp

```diff
@@ -310,3 +297,10 @@ NodeBase::resolve_topic_or_service_name(
   allocator.deallocate(output_cstr, allocator.state);
   return output;
 }
+
+void
+NodeBase::trigger_notify_guard_condition()
+{
+  std::unique_lock<std::recursive_mutex> lock(notify_guard_condition_mutex_);
+  notify_guard_condition_.trigger();
+}
```

Comment:
With the introduction of this function, maybe we don't need the `acquire_notify_guard_condition_lock` any longer and should deprecate it.

I can't find why we had it this way, it seemed intended to be used like this:

```c++
  {
    auto notify_condition_lock = node_base_->acquire_notify_guard_condition_lock();
    // some other "critical section" stuff
    rcl_ret_t ret = rcl_trigger_guard_condition(node_base_->get_notify_guard_condition());
    if (RCL_RET_OK != ret) {
      throw_from_rcl_error(ret, "failed to trigger notify guard condition");
    }
  }
```

But it doesn't appear to be used like that anywhere. At least not anymore.

And now I'm wondering if the lock is needed at all. I think triggering a guard condition should be thread-safe...

Reply:
> With the introduction of this function, maybe we don't need the acquire_notify_guard_condition_lock any longer and should deprecate it.

> I'm wondering if the lock is needed at all. I think triggering a guard condition should be thread-safe

I also don't think the lock is needed, as triggering the guard condition is thread-safe.
Done in 82629b7